x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/optimizer/Optimizer.java

```diff
@@ -1148,29 +1148,40 @@ private static LogicalPlan skipPlan(UnaryPlan plan) {
     static class SkipQueryIfFoldingProjection extends OptimizerRule<LogicalPlan> {
         @Override
         protected LogicalPlan rule(LogicalPlan plan) {
-            Holder<LocalRelation> optimizedPlan = new Holder<>();
-            plan.forEachDown(Project.class, p -> {
-                List<Object> values = extractConstants(p.projections());
-                if (values.size() == p.projections().size() && (p.child() instanceof EsRelation) == false &&
-                    isNotQueryWithFromClauseAndFilterFoldedToFalse(p)) {
-                    optimizedPlan.set(new LocalRelation(p.source(), new SingletonExecutable(p.output(), values.toArray())));
-                }
-            });
+            List<LogicalPlan> leaves = plan.collectLeaves();
 
-            if (optimizedPlan.get() != null) {
-                return optimizedPlan.get();
-            }
+            List<LogicalPlan> projectOrAggregates = plan.collect(p ->
+                p instanceof Project || p instanceof Aggregate);
+
+            if (leaves.size() == 1 && projectOrAggregates.size() == 1) {
+                LogicalPlan leaf = leaves.get(0);
+                LogicalPlan projectOrAggregate = projectOrAggregates.get(0);
+
+                List<Object> foldedValues = null;
+
+                // exclude LocalRelations that have been introduced by earlier optimizations (skipped ESRelations)
+                boolean isNonSkippedLocalRelation = leaf instanceof LocalRelation
+                    && ((LocalRelation) leaf).executable() instanceof EmptyExecutable == false;
 
-            plan.forEachDown(Aggregate.class, a -> {
-                List<Object> values = extractConstants(a.aggregates());
-                if (values.size() == a.aggregates().size() && a.groupings().isEmpty()
-                    && isNotQueryWithFromClauseAndFilterFoldedToFalse(a)) {
-                    optimizedPlan.set(new LocalRelation(a.source(), new SingletonExecutable(a.output(), values.toArray())));
+                if (projectOrAggregate instanceof Project && isNonSkippedLocalRelation) {
```

Comment:
The if/else could be optimized by swapping it around and avoid checking for Project:
```
if (projectOrAggregate instanceof Aggregate) {
 ...
 }
else if (isNonSkippedLocalRelation) // no need to check for Project
```